### PR TITLE
[ep] Avoid mooncake ep test crash when ibgda_init fails

### PR DESCRIPTION
**Pull request by zxpdemonio**
## Description

Make Mooncake-EP robust when IBGDA init fails (e.g., eRDMA / no GPUDirect RDMA) by safely entering fallback mode and preventing test/profiling crashes.

## Type of Change

* Types
  - [ ] Bug fix
  - [ ] New feature
    - [ ] Transfer Engine
    - [ ] Mooncake Store
    - [ ] Mooncake EP
    - [ ] Integration
    - [ ] P2P Store
    - [ ] Python Wheel
  - [ ] Breaking change
  - [ ] CI/CD
  - [ ] Documentation update
  - [ ] Other

## Checklist

- [ ] I have performed a self-review of my own code.
- [ ] I have formatted my own code using `./scripts/code_format.sh` before submitting.
- [ ] I have updated the documentation.
- [ ] I have added tests to prove my changes are effective.

### CODE REVIEW

File: mooncake-wheel/mooncake/mooncake_ep_buffer.py [hunk 1]
```diff
@@ -120,16 +120,34 @@ def __init__(self, group: dist.ProcessGroup, num_ep_buffer_bytes: int = 0):
...
-        # pybind11 converts std::vector<int32_t> to a list of integers
-        # Convert list to tensor
-        local_handle_tensor = torch.tensor(local_handle_ints, dtype=torch.int32, device='cuda')
-        handles = [torch.empty(len(local_handle_ints), dtype=torch.int32, device='cuda') for _ in range(self.group_size)]
-        dist.all_gather(handles, local_handle_tensor, group)
-        remote_handles = [h.tolist() for h in handles]
-        self.runtime.sync_nvlink_ipc_handles(remote_handles)
-        self._use_fallback = bool(self.runtime.ibgda_disabled())
+        # Exchange CUDA IPC handles for NVLink P2P (only in non-fallback mode)
+        if not self._use_fallback:
```
Comment (UNIDY2002) [inline mooncake-wheel/mooncake/mooncake_ep_buffer.py]: Actually, we *must* exchange IPC handles for NVLink in fallback mode, considering the SGLang CI environment where NVLink is available while IBGDA is not. I would recommend removing the condition, since the try...catch block already handles the IPC-handle-exchange-failure correctly.

File: mooncake-wheel/mooncake/mooncake_ep_buffer.py [hunk 2]
```diff
@@ -120,16 +120,34 @@ def __init__(self, group: dist.ProcessGroup, num_ep_buffer_bytes: int = 0):
...
+                import warnings
+                warnings.warn(
+                    f"[Rank {self.rank}] Failed to exchange IPC handles: {e}. Falling back.",
+                    RuntimeWarning,
+                    stacklevel=2,
+                )
+                self._use_fallback = True
+
+        # Re-check fallback status after IPC handle exchange
+        self._use_fallback = self._use_fallback or bool(self.runtime.ibgda_disabled())
```
Comment (UNIDY2002) [inline mooncake-wheel/mooncake/mooncake_ep_buffer.py]:
```suggestion
```

Move the original line inside the try block:
```
self._use_fallback = bool(self.runtime.ibgda_disabled())
```

File: mooncake-wheel/mooncake/mooncake_ep_buffer.py [hunk 3]
```diff
@@ -120,16 +121,33 @@ def __init__(self, group: dist.ProcessGroup, num_ep_buffer_bytes: int = 0):
...
+            warnings.warn(
+                f"[Rank {self.rank}] Failed to exchange IPC handles: {e}. Falling back.",
+                RuntimeWarning,
+                stacklevel=2,
+            )
+            self._use_fallback = True
+
+        # Re-check fallback status after IPC handle exchange
+        # (cache `ibgda_disabled` to avoid calling into the runtime twice)
+        self._use_fallback = self._use_fallback or ibgda_disabled
```
Comment (UNIDY2002) [inline mooncake-wheel/mooncake/mooncake_ep_buffer.py]: The value of `self.runtime.ibgda_disabled()` may change after calling `self.runtime.sync_nvlink_ipc_handles()`, so we cannot cache the value in advance.

```suggestion
```

File: mooncake-wheel/mooncake/mooncake_ep_buffer.py [hunk 4]
```diff
@@ -120,16 +121,33 @@ def __init__(self, group: dist.ProcessGroup, num_ep_buffer_bytes: int = 0):
...
+        # - If this fails (no NVLink / CUDA IPC unavailable / platform restrictions), we swallow the
+        #   error and keep going in fallback mode.
+        try:
+            local_handle_ints = self.runtime.get_ipc_handle()
+            # pybind11 converts std::vector<int32_t> to a list of integers
+            local_handle_tensor = torch.tensor(local_handle_ints, dtype=torch.int32, device='cuda')
+            handles = [torch.empty(len(local_handle_ints), dtype=torch.int32, device='cuda') for _ in range(self.group_size)]
+            dist.all_gather(handles, local_handle_tensor, group)
+            remote_handles = [h.tolist() for h in handles]
+            self.runtime.sync_nvlink_ipc_handles(remote_handles)
```
Comment (UNIDY2002) [inline mooncake-wheel/mooncake/mooncake_ep_buffer.py:139]:
```suggestion
            self.runtime.sync_nvlink_ipc_handles(remote_handles)
            self._use_fallback = bool(self.runtime.ibgda_disabled())
```

File: mooncake-wheel/mooncake/mooncake_ep_buffer.py [hunk 5]
```diff
@@ -72,7 +72,8 @@ def __init__(self, group: dist.ProcessGroup, num_ep_buffer_bytes: int = 0):
         preferred_hca = pg.get_preferred_hca(self.backend, f'cuda:{torch.cuda.current_device()}')
         self.runtime = ep.Buffer(self.rank, self.group_size, num_ep_buffer_bytes, preferred_hca)
         # Fallback flag and buffers
-        self._use_fallback = bool(self.runtime.ibgda_disabled())
+        ibgda_disabled = bool(self.runtime.ibgda_disabled())
+        self._use_fallback = ibgda_disabled
```
Comment (UNIDY2002) [inline mooncake-wheel/mooncake/mooncake_ep_buffer.py]:
```suggestion
        self._use_fallback = bool(self.runtime.ibgda_disabled())
```

**Review by UNIDY2002 — COMMENTED** [review]
Mostly good, with one thing. There is an undocumented behavior of `sync_nvlink_ipc_handles()` that may set `ibgda_disabled` to false during certain conditions.

(I admit that the variable naming sounds counterintuitive. But let's leave the variable naming issue to future work, as it requires refactoring of the C++ part.)

**Comment by zxpdemonio** [comment]
> Mostly good, with one thing. There is an undocumented behavior of `sync_nvlink_ipc_handles()` that may set `ibgda_disabled` to false during certain conditions.
> 
> (I admit that the variable naming sounds counterintuitive. But let's leave the variable naming issue to future work, as it requires refactoring of the C++ part.)

I've updated the PR with additional comments explaining how ibgda_disabled might be toggled within sync_nvlink_ipc_handles()

**Comment by zxpdemonio** [comment]
Wait a minute, there's still problem with fallback mode, I'll check and fix it.

**Comment by stmatengss** [comment]
> Wait a minute, there's still problem with fallback mode, I'll check and fix it.

Could you update this PR?

File: mooncake-ep/include/mooncake_ep_buffer.h [hunk 6]
```diff
@@ -126,6 +127,19 @@ struct MooncakeEpBuffer {
...
+    // instead of Python fallback implementation.
+    // Returns true if either IBGDA is available OR P2P+IPC is fully enabled
+    // AND IBGDA resources are initialized (even if IBGDA is disabled).
+    bool use_fast_path() {
+        if (!ibgda_disabled_) {
+            return true;  // IBGDA available, safe to use fast-path
+        }
+        // IBGDA disabled, check if P2P+IPC is fully enabled
+        // AND IBGDA resources are initialized (for fallback in kernel)
+        return p2p_ipc_all_enabled_ && mr != nullptr && !qps.empty();
```
Comment (UNIDY2002) [inline mooncake-ep/include/mooncake_ep_buffer.h]: As long as P2P IPC (NVLink) is all enabled, the EP kernels can run successfully even if qps is empty. The original line is a bit conservative.

```suggestion
        return p2p_ipc_all_enabled_;
```

File: mooncake-wheel/mooncake/mooncake_ep_buffer.py [hunk 7]
```diff
@@ -120,16 +121,49 @@ def __init__(self, group: dist.ProcessGroup, num_ep_buffer_bytes: int = 0):
...
+            )
+
+        # Final decision: Use fast-path (CUDA kernel + IBGDA/NVLink) only if it's safe.
+        # The runtime checks:
+        # - If IBGDA is available, use fast-path
+        # - If IBGDA is unavailable but P2P+IPC is fully enabled AND IBGDA resources
+        #   are initialized (for fallback in kernel), use fast-path
+        # - Otherwise, fall back to Python implementation
+        ibgda_disabled = bool(self.runtime.ibgda_disabled())
+        p2p_ipc_all_enabled = False
```
Comment (UNIDY2002) [inline mooncake-wheel/mooncake/mooncake_ep_buffer.py]: The variable `p2p_ipc_all_enabled` seems to be unused.

File: mooncake-ep/include/mooncake_ep_buffer.h (hunk 6 above)
Comment (UNIDY2002) [inline mooncake-ep/include/mooncake_ep_buffer.h]: Please update this line. Thank you!

```suggestion
        return p2p_ipc_all_enabled_;
```

Comment (stmatengss) [inline mooncake-ep/include/mooncake_ep_buffer.h]: Hi @zxpdemonio , you don't update this line.

Comment (zxpdemonio) [reply mooncake-ep/include/mooncake_ep_buffer.h]: Actually, the test will fail in erdma environment if we remove "&& mr != nullptr && !qps.empty()", and I will investigate the root cause further.

**Review by UNIDY2002 — CHANGES_REQUESTED** [review]
Wait a minute. The accuracy is affected in e2e tests.

**Comment by UNIDY2002** [comment]
BTW, the accuracy issue arise from the main branch, so it is safe here.
